### PR TITLE
count all queries when loading a hoc level page

### PR DESCRIPTION
**Pull request by davidsbailey**
Finishes https://codedotorg.atlassian.net/browse/PLAT-1181. Mostly copies https://github.com/code-dot-org/code-dot-org/pull/42737. 

This PR just adds a new unit test covering all the requests I could capture when loading a script level page. The first thing to note is that there are no cacheable curriculum-related objects being queried from the database, which was the explicit goal of PLAT-1181. I also poked around a bit to see if I could shave off any other queries, but didn't see any low hanging fruit. So, my proposal is to add this query count test, in hopes of (1) documenting some of the existing queries on the uncached HOC script level page, and (2) protecting against regressions to this query count.

## Testing story

* new unit test to count queries
* relying on existing tests to cover the small non-test code change

## Future work

If we were going to try to shave off more queries here in the future, I would think about doing it by disabling certain features for certain high profile scripts, including:
* teacher feedback
* `section_hidden_scripts` (does this even make sense for scripts which are standalone courses?)
* `section_hidden_stages` (does this even make sense for scripts with one lesson?)
* rubrics

of course, adding any HOC scripts to the [CACHED_UNITS_MAP](https://github.com/code-dot-org/code-dot-org/blob/8f6785bb50e7b8994bec321271b3f9d82d8b80fa/cookbooks/cdo-varnish/libraries/http_cache.rb#L28) would strip the user session cookies from the requests and result in avoiding many (most? almost all?) of these queries.

### CODE REVIEW

File: dashboard/test/integration/db_query_test.rb [hunk 1]
```diff
@@ -127,4 +127,43 @@ def setup
...
+    section = create :section, user: teacher, script: unit
+    student = create :student
+    section.students = [student]
+    student.assign_script(unit)
+    sign_in student
+
+    assert_cached_queries(19) do
+      get "/s/#{unit.name}/lessons/1/levels/1"
+      assert_response :success
+    end
```
Comment (davidsbailey) [inline dashboard/test/integration/db_query_test.rb:154]: These are the existing queries, which I got by temporarily changing the query count to zero, and then filtering out some irrelevant parts of the stack traces:
```
FAIL["test_student_in_section_views_uncached_hoc_level", "DBQueryTest", 29.956948000006378]
test_student_in_section_views_uncached_hoc_level#DBQueryTest (29.96s)
Wrong query count:

  User Load (0.0ms)  SELECT  `users`.* FROM `users` WHERE `users`.`deleted_at` IS NULL AND `users`.`id` = 55 ORDER BY `users`.`id` ASC LIMIT 1
app/controllers/application_controller.rb:224:in `current_user'
app/controllers/application_controller.rb:264:in `redirect_admin_from_labs'

  Follower Load (0.0ms)  SELECT `followers`.* FROM `followers` WHERE `followers`.`deleted_at` IS NULL AND `followers`.`student_user_id` = 55 ORDER BY followers.id
app/controllers/script_levels_controller.rb:80:in `show'

  Section Load (0.0ms)  SELECT `sections`.* FROM `sections` WHERE `sections`.`deleted_at` IS NULL AND `sections`.`id` = 40
app/controllers/script_levels_controller.rb:80:in `show'

  User Load (0.0ms)  SELECT `users`.* FROM `users` WHERE `users`.`deleted_at` IS NULL AND `users`.`id` = 54
app/controllers/script_levels_controller.rb:80:in `show'

  Section Load (0.0ms)  SELECT `sections`.* FROM `sections` INNER JOIN `followers` ON `sections`.`id` = `followers`.`section_id` WHERE `sections`.`deleted_at` IS NULL AND `followers`.`deleted_at` IS NULL AND `followers`.`student_user_id` = 55 AND `sections`.`hidden` = FALSE ORDER BY followers.id
app/controllers/script_levels_controller.rb:114:in `map'
app/controllers/script_levels_controller.rb:114:in `show'

  SectionHiddenLesson Load (0.0ms)  SELECT  `section_hidden_stages`.* FROM `section_hidden_stages` WHERE `section_hidden_stages`.`stage_id` = 255 AND `section_hidden_stages`.`section_id` = 40 LIMIT 1
app/models/script_level.rb:623:in `hidden_for_section?'
app/models/user.rb:1267:in `block in script_level_hidden?'
app/models/user.rb:1267:in `each'
app/models/user.rb:1267:in `all?'
app/models/user.rb:1267:in `script_level_hidden?'
app/controllers/script_levels_controller.rb:122:in `show'

  SectionHiddenScript Load (0.0ms)  SELECT  `section_hidden_scripts`.* FROM `section_hidden_scripts` WHERE `section_hidden_scripts`.`script_id` = 96 AND `section_hidden_scripts`.`section_id` = 40 LIMIT 1
app/models/script_level.rb:624:in `hidden_for_section?'
app/models/user.rb:1267:in `block in script_level_hidden?'
app/models/user.rb:1267:in `each'
app/models/user.rb:1267:in `all?'
app/models/user.rb:1267:in `script_level_hidden?'
app/controllers/script_levels_controller.rb:122:in `show'

   (0.0ms)  SELECT COUNT(*) FROM `sections` WHERE `sections`.`deleted_at` IS NULL AND `sections`.`user_id` = 55 AND `sections`.`hidden` = FALSE
app/controllers/script_levels_controller.rb:430:in `load_section'
app/controllers/script_levels_controller.rb:160:in `show'

  UserLevel Load (0.0ms)  SELECT  `user_levels`.* FROM `user_levels` WHERE `user_levels`.`deleted_at` IS NULL AND `user_levels`.`user_id` = 55 AND `user_levels`.`level_id` = 333 AND `user_levels`.`script_id` = 96 ORDER BY updated_at DESC LIMIT 1
app/models/user.rb:1237:in `last_attempt'
app/controllers/script_levels_controller.rb:377:in `load_level_source'
app/controllers/script_levels_controller.rb:486:in `present_level'
app/controllers/script_levels_controller.rb:167:in `show'

  UserLevel Load (0.0ms)  SELECT  `user_levels`.* FROM `user_levels` WHERE `user_levels`.`deleted_at` IS NULL AND `user_levels`.`user_id` = 55 AND `user_levels`.`script_id` = 96 AND `user_levels`.`level_id` = 333 LIMIT 1
app/controllers/script_levels_controller.rb:380:in `load_level_source'
app/controllers/script_levels_controller.rb:486:in `present_level'
app/controllers/script_levels_controller.rb:167:in `show'

  SectionHiddenLesson Load (0.0ms)  SELECT  `section_hidden_stages`.* FROM `section_hidden_stages` WHERE `section_hidden_stages`.`stage_id` = 256 AND `section_hidden_stages`.`section_id` = 40 LIMIT 1
app/models/script_level.rb:623:in `hidden_for_section?'
app/models/user.rb:1267:in `block in script_level_hidden?'
app/models/user.rb:1267:in `each'
app/models/user.rb:1267:in `all?'
app/models/user.rb:1267:in `script_level_hidden?'
app/models/script_level.rb:295:in `locked_or_hidden?'
app/models/script_level.rb:289:in `valid_progression_level?'
app/models/script_level.rb:280:in `or_next_progression_level'
app/models/script_level.rb:274:in `next_progression_level'
app/models/script_level.rb:235:in `next_level_or_redirect_path_for_user'
app/helpers/levels_helper.rb:208:in `app_options'
app/views/levels/_blockly.html.haml:19:in `_app_views_levels__blockly_html_haml__1900486341189874240_70185011498300'
app/views/levels/show.html.haml:63:in `_app_views_levels_show_html_haml___4466479542367776872_70185016223380'
app/controllers/script_levels_controller.rb:544:in `present_level'
app/controllers/script_levels_controller.rb:167:in `show'

   (0.0ms)  SELECT `authored_hint_view_requests`.`hint_id` FROM `authored_hint_view_requests` WHERE `authored_hint_view_requests`.`user_id` = 55 AND `authored_hint_view_requests`.`script_id` = 96 AND `authored_hint_view_requests`.`level_id` = 333
app/helpers/levels_helper.rb:242:in `app_options'
app/views/levels/_blockly.html.haml:19:in `_app_views_levels__blockly_html_haml__1900486341189874240_70185011498300'
app/views/levels/show.html.haml:63:in `_app_views_levels_show_html_haml___4466479542367776872_70185016223380'
app/controllers/script_levels_controller.rb:544:in `present_level'
app/controllers/script_levels_controller.rb:167:in `show'

   (0.0ms)  SELECT `user_permissions`.`permission` FROM `user_permissions` WHERE `user_permissions`.`user_id` = 54
app/models/concerns/user_permission_grantee.rb:22:in `permission?'
app/models/user.rb:1327:in `authorized_teacher?'
app/models/user.rb:1336:in `student_of_authorized_teacher?'
app/helpers/users_helper.rb:117:in `user_summary'
app/helpers/users_helper.rb:83:in `summarize_user_progress'
app/views/layouts/_header.html.haml:101:in `_app_views_layouts__header_html_haml___545151121208187762_70185061968800'
app/views/layouts/application.html.haml:68:in `_app_views_layouts_application_html_haml___1175173215306511820_70185016322820'
app/controllers/script_levels_controller.rb:544:in `present_level'
app/controllers/script_levels_controller.rb:167:in `show'

  UserLevel Load (0.0ms)  SELECT `user_levels`.* FROM `user_levels` WHERE `user_levels`.`deleted_at` IS NULL AND `user_levels`.`user_id` = 55 AND `user_levels`.`script_id` = 96 ORDER BY id desc
app/models/user.rb:1085:in `user_levels_by_level'
app/helpers/users_helper.rb:205:in `merge_script_progress'
app/helpers/users_helper.rb:84:in `summarize_user_progress'
app/views/layouts/_header.html.haml:101:in `_app_views_layouts__header_html_haml___545151121208187762_70185061968800'
app/views/layouts/application.html.haml:68:in `_app_views_layouts_application_html_haml___1175173215306511820_70185016322820'
app/controllers/script_levels_controller.rb:544:in `present_level'
app/controllers/script_levels_controller.rb:167:in `show'

   (0.0ms)  SELECT MAX(teacher_feedbacks.id) FROM `teacher_feedbacks` INNER JOIN `users` ON `users`.`deleted_at` IS NULL AND `users`.`id` = `teacher_feedbacks`.`student_id` INNER JOIN `followers` ON `followers`.`deleted_at` IS NULL AND `followers`.`student_user_id` = `users`.`id` INNER JOIN `sections` ON `sections`.`deleted_at` IS NULL AND `sections`.`id` = `followers`.`section_id` WHERE `teacher_feedbacks`.`deleted_at` IS NULL AND `teacher_feedbacks`.`student_id` = 55 AND `teacher_feedbacks`.`script_id` = 96 AND (sections.user_id = teacher_id) GROUP BY `teacher_feedbacks`.`teacher_id`, `teacher_feedbacks`.`student_id`, `teacher_feedbacks`.`level_id`
app/models/teacher_feedback.rb:109:in `latest_per_teacher'
app/models/teacher_feedback.rb:77:in `get_latest_feedbacks_received'
app/helpers/users_helper.rb:183:in `teacher_feedbacks_by_student_by_level'
app/helpers/users_helper.rb:206:in `merge_script_progress'
app/helpers/users_helper.rb:84:in `summarize_user_progress'
app/views/layouts/_header.html.haml:101:in `_app_views_layouts__header_html_haml___545151121208187762_70185061968800'
app/views/layouts/application.html.haml:68:in `_app_views_layouts_application_html_haml___1175173215306511820_70185016322820'
app/controllers/script_levels_controller.rb:544:in `present_level'
app/controllers/script_levels_controller.rb:167:in `show'

  TeacherFeedback Load (0.0ms)  SELECT `teacher_feedbacks`.* FROM `teacher_feedbacks` WHERE `teacher_feedbacks`.`deleted_at` IS NULL AND `teacher_feedbacks`.`student_id` = 55 AND `teacher_feedbacks`.`script_id` = 96 AND 1=0 ORDER BY `teacher_feedbacks`.`created_at` DESC
app/helpers/users_helper.rb:184:in `group_by'
app/helpers/users_helper.rb:184:in `teacher_feedbacks_by_student_by_level'
app/helpers/users_helper.rb:206:in `merge_script_progress'
app/helpers/users_helper.rb:84:in `summarize_user_progress'
app/views/layouts/_header.html.haml:101:in `_app_views_layouts__header_html_haml___545151121208187762_70185061968800'
app/views/layouts/application.html.haml:68:in `_app_views_layouts_application_html_haml___1175173215306511820_70185016322820'
app/controllers/script_levels_controller.rb:544:in `present_level'
app/controllers/script_levels_controller.rb:167:in `show'

   (0.0ms)  SELECT `paired_user_levels`.`driver_user_level_id` FROM `paired_user_levels` WHERE 1=0
app/models/paired_user_level.rb:34:in `pairs'
app/helpers/users_helper.rb:207:in `merge_script_progress'
app/helpers/users_helper.rb:84:in `summarize_user_progress'
app/views/layouts/_header.html.haml:101:in `_app_views_layouts__header_html_haml___545151121208187762_70185061968800'
app/views/layouts/application.html.haml:68:in `_app_views_layouts_application_html_haml___1175173215306511820_70185016322820'
app/controllers/script_levels_controller.rb:544:in `present_level'
app/controllers/script_levels_controller.rb:167:in `show'

   (0.0ms)  SELECT `paired_user_levels`.`navigator_user_level_id` FROM `paired_user_levels` WHERE 1=0
app/models/paired_user_level.rb:37:in `pairs'
app/helpers/users_helper.rb:207:in `merge_script_progress'
app/helpers/users_helper.rb:84:in `summarize_user_progress'
app/views/layouts/_header.html.haml:101:in `_app_views_layouts__header_html_haml___545151121208187762_70185061968800'
app/views/layouts/application.html.haml:68:in `_app_views_layouts_application_html_haml___1175173215306511820_70185016322820'
app/controllers/script_levels_controller.rb:544:in `present_level'
app/controllers/script_levels_controller.rb:167:in `show'

  UserScript Load (0.0ms)  SELECT  `user_scripts`.* FROM `user_scripts` WHERE `user_scripts`.`deleted_at` IS NULL AND `user_scripts`.`user_id` = 55 AND `user_scripts`.`script_id` = 96 ORDER BY -completed_at asc, greatest(coalesce(started_at, 0), coalesce(assigned_at, 0), coalesce(last_progress_at, 0)) desc, user_scripts.id asc LIMIT 1
lib/policies/script_activity.rb:5:in `completed?'
app/helpers/users_helper.rb:208:in `merge_script_progress'
app/helpers/users_helper.rb:84:in `summarize_user_progress'
app/views/layouts/_header.html.haml:101:in `_app_views_layouts__header_html_haml___545151121208187762_70185061968800'
app/views/layouts/application.html.haml:68:in `_app_views_layouts_application_html_haml___1175173215306511820_70185016322820'
app/controllers/script_levels_controller.rb:544:in `present_level'
app/controllers/script_levels_controller.rb:167:in `show'
.
Expected: 0
  Actual: 19
```

File: dashboard/test/integration/db_query_test.rb [hunk 2]
```diff
@@ -127,4 +127,43 @@ def setup
...
+
+    # Simulate all the ajax requests which the level page sends to the
+    # server on page load.
+
+    assert_cached_queries(10) do
+      get "/api/user_progress/#{unit.name}/1/1/#{level.id}"
+      assert_response :success
+    end
```
Comment (davidsbailey) [inline dashboard/test/integration/db_query_test.rb:162]:
```
 FAIL["test_student_in_section_views_uncached_hoc_level", "DBQueryTest", 30.483966999920085]
 test_student_in_section_views_uncached_hoc_level#DBQueryTest (30.48s)
Wrong query count:

  User Load (0.0ms)  SELECT  `users`.* FROM `users` WHERE `users`.`deleted_at` IS NULL AND `users`.`id` = 57 ORDER BY `users`.`id` ASC LIMIT 1
app/controllers/application_controller.rb:224:in `current_user'
app/controllers/api_controller.rb:421:in `user_progress_for_lesson'

  User Load (0.0ms)  SELECT `users`.* FROM `users` INNER JOIN `sections` ON `users`.`id` = `sections`.`user_id` INNER JOIN `followers` ON `sections`.`id` = `followers`.`section_id` WHERE `users`.`deleted_at` IS NULL AND `sections`.`deleted_at` IS NULL AND `followers`.`deleted_at` IS NULL AND `followers`.`student_user_id` = 57 ORDER BY followers.id
app/models/user.rb:1336:in `student_of_authorized_teacher?'
app/helpers/users_helper.rb:117:in `user_summary'
app/controllers/api_controller.rb:421:in `user_progress_for_lesson'

   (0.0ms)  SELECT `user_permissions`.`permission` FROM `user_permissions` WHERE `user_permissions`.`user_id` = 56
app/models/concerns/user_permission_grantee.rb:22:in `permission?'
app/models/user.rb:1327:in `authorized_teacher?'
app/models/user.rb:1336:in `student_of_authorized_teacher?'
app/helpers/users_helper.rb:117:in `user_summary'
app/controllers/api_controller.rb:421:in `user_progress_for_lesson'

  UserLevel Load (0.0ms)  SELECT  `user_levels`.* FROM `user_levels` WHERE `user_levels`.`deleted_at` IS NULL AND `user_levels`.`user_id` = 57 AND `user_levels`.`level_id` = 383 AND `user_levels`.`script_id` = 105 ORDER BY updated_at DESC LIMIT 1
app/models/user.rb:1237:in `last_attempt'
app/controllers/api_controller.rb:430:in `user_progress_for_lesson'

  UserLevel Load (0.0ms)  SELECT `user_levels`.* FROM `user_levels` WHERE `user_levels`.`deleted_at` IS NULL AND `user_levels`.`user_id` = 57 AND `user_levels`.`script_id` = 105 ORDER BY id desc
app/models/user.rb:1085:in `user_levels_by_level'
app/helpers/users_helper.rb:205:in `merge_script_progress'
app/helpers/users_helper.rb:84:in `summarize_user_progress'
app/controllers/api_controller.rb:434:in `user_progress_for_lesson'

   (0.0ms)  SELECT MAX(teacher_feedbacks.id) FROM `teacher_feedbacks` INNER JOIN `users` ON `users`.`deleted_at` IS NULL AND `users`.`id` = `teacher_feedbacks`.`student_id` INNER JOIN `followers` ON `followers`.`deleted_at` IS NULL AND `followers`.`student_user_id` = `users`.`id` INNER JOIN `sections` ON `sections`.`deleted_at` IS NULL AND `sections`.`id` = `followers`.`section_id` WHERE `teacher_feedbacks`.`deleted_at` IS NULL AND `teacher_feedbacks`.`student_id` = 57 AND `teacher_feedbacks`.`script_id` = 105 AND (sections.user_id = teacher_id) GROUP BY `teacher_feedbacks`.`teacher_id`, `teacher_feedbacks`.`student_id`, `teacher_feedbacks`.`level_id`
app/models/teacher_feedback.rb:109:in `latest_per_teacher'
app/models/teacher_feedback.rb:77:in `get_latest_feedbacks_received'
app/helpers/users_helper.rb:183:in `teacher_feedbacks_by_student_by_level'
app/helpers/users_helper.rb:206:in `merge_script_progress'
app/helpers/users_helper.rb:84:in `summarize_user_progress'
app/controllers/api_controller.rb:434:in `user_progress_for_lesson'

  TeacherFeedback Load (0.0ms)  SELECT `teacher_feedbacks`.* FROM `teacher_feedbacks` WHERE `teacher_feedbacks`.`deleted_at` IS NULL AND `teacher_feedbacks`.`student_id` = 57 AND `teacher_feedbacks`.`script_id` = 105 AND 1=0 ORDER BY `teacher_feedbacks`.`created_at` DESC
app/helpers/users_helper.rb:184:in `group_by'
app/helpers/users_helper.rb:184:in `teacher_feedbacks_by_student_by_level'
app/helpers/users_helper.rb:206:in `merge_script_progress'
app/helpers/users_helper.rb:84:in `summarize_user_progress'
app/controllers/api_controller.rb:434:in `user_progress_for_lesson'

   (0.0ms)  SELECT `paired_user_levels`.`driver_user_level_id` FROM `paired_user_levels` WHERE 1=0
app/models/paired_user_level.rb:34:in `pairs'
app/helpers/users_helper.rb:207:in `merge_script_progress'
app/helpers/users_helper.rb:84:in `summarize_user_progress'
app/controllers/api_controller.rb:434:in `user_progress_for_lesson'

   (0.0ms)  SELECT `paired_user_levels`.`navigator_user_level_id` FROM `paired_user_levels` WHERE 1=0
app/models/paired_user_level.rb:37:in `pairs'
app/helpers/users_helper.rb:207:in `merge_script_progress'
app/helpers/users_helper.rb:84:in `summarize_user_progress'
app/controllers/api_controller.rb:434:in `user_progress_for_lesson'

  UserScript Load (0.0ms)  SELECT  `user_scripts`.* FROM `user_scripts` WHERE `user_scripts`.`deleted_at` IS NULL AND `user_scripts`.`user_id` = 57 AND `user_scripts`.`script_id` = 105 ORDER BY -completed_at asc, greatest(coalesce(started_at, 0), coalesce(assigned_at, 0), coalesce(last_progress_at, 0)) desc, user_scripts.id asc LIMIT 1
lib/policies/script_activity.rb:5:in `completed?'
app/helpers/users_helper.rb:208:in `merge_script_progress'
app/helpers/users_helper.rb:84:in `summarize_user_progress'
app/controllers/api_controller.rb:434:in `user_progress_for_lesson'
.
Expected: 0
  Actual: 10
```

File: dashboard/test/integration/db_query_test.rb [hunk 3]
```diff
@@ -127,4 +127,43 @@ def setup
...
+
+    assert_cached_queries(3) do
+      get "/levels/#{level.id}/get_rubric"
+      assert_response :success
+    end
```
Comment (davidsbailey) [inline dashboard/test/integration/db_query_test.rb:167]:
```
 FAIL["test_student_in_section_views_uncached_hoc_level", "DBQueryTest", 29.04486500006169]
 test_student_in_section_views_uncached_hoc_level#DBQueryTest (29.04s)
        Wrong query count:

  Level Load (0.0ms)  SELECT  `levels`.* FROM `levels` WHERE `levels`.`id` = 433 LIMIT 1
config/initializers/mysql_check_index_used.rb:22:in `execute'
app/helpers/read_replica_helper.rb:20:in `set_read_only_connection_for_block'
test/integration/db_query_test.rb:165:in `block (2 levels) in <class:DBQueryTest>'
test/integration/db_query_test.rb:164:in `block in <class:DBQueryTest>'

  User Load (0.0ms)  SELECT  `users`.* FROM `users` WHERE `users`.`deleted_at` IS NULL AND `users`.`id` = 59 ORDER BY `users`.`id` ASC LIMIT 1
config/initializers/mysql_check_index_used.rb:22:in `execute'
app/controllers/application_controller.rb:224:in `current_user'
app/helpers/read_replica_helper.rb:20:in `set_read_only_connection_for_block'
test/integration/db_query_test.rb:165:in `block (2 levels) in <class:DBQueryTest>'
test/integration/db_query_test.rb:164:in `block in <class:DBQueryTest>'

  User Load (0.0ms)  SELECT `users`.* FROM `users` INNER JOIN `sections` ON `users`.`id` = `sections`.`user_id` INNER JOIN `followers` ON `sections`.`id` = `followers`.`section_id` WHERE `users`.`deleted_at` IS NULL AND `sections`.`deleted_at` IS NULL AND `followers`.`deleted_at` IS NULL AND `followers`.`student_user_id` = 59 ORDER BY followers.id
config/initializers/mysql_check_index_used.rb:22:in `execute'
app/models/ability.rb:370:in `initialize'
app/helpers/read_replica_helper.rb:20:in `set_read_only_connection_for_block'
test/integration/db_query_test.rb:165:in `block (2 levels) in <class:DBQueryTest>'
test/integration/db_query_test.rb:164:in `block in <class:DBQueryTest>'
.
Expected: 0
  Actual: 3
```

**Review by jamescodeorg — APPROVED** [review]
Thanks for the analysis and new tests!

**Pull request merged**